protected/humhub/helpers/DataTypeHelper.php

```diff
@@ -0,0 +1,628 @@
+<?php
+
+/*
+ * @link      https://www.humhub.org/
+ * @copyright Copyright (c) 2023 HumHub GmbH & Co. KG
+ * @license   https://www.humhub.com/licences
+ */
+
+namespace humhub\helpers;
+
+use humhub\exceptions\InvalidArgumentClassException;
+use humhub\exceptions\InvalidArgumentTypeException;
+use humhub\exceptions\InvalidArgumentValueException;
+use LogicException;
+use Stringable;
+
+/**
+ * @since 1.16
+ */
+class DataTypeHelper
+{
+    public const CLASS_CHECK_VALUE_IS_NULL = 128;
+    public const CLASS_CHECK_TYPE_NOT_IN_LIST = 32;
+    public const CLASS_CHECK_INVALID_TYPE_PARAMETER = 2;
+    public const CLASS_CHECK_VALUE_IS_EMPTY = 4;
+    public const CLASS_CHECK_INVALID_CLASSNAME_PARAMETER = 1;
+    public const CLASS_CHECK_INVALID_TYPE = 8;
+    public const CLASS_CHECK_NON_EXISTING_CLASS = 16;
+    public const CLASS_CHECK_VALUE_IS_INSTANCE = 64;
+
+
+    /**
+     * @param mixed $input
+     * @param string|string[] $types
+     * @param bool|null $requireAll
+     * @param string|null $throw = Name of the argument to be used for InvalidArgumentTypeException or Null if no
+     *     exception should be thrown
+     * @param array|null $typesChecked
+     *
+     * @return string|null
+     * @since 1.16
+     */
+    public static function checkType($input, $types, ?bool $requireAll = false, ?string $throw = '$input', ?array &$typesChecked = null): ?string
```

Comment:
I would prefer the method to be private.

```suggestion
    private static function checkType($input, $types, ?bool $requireAll = false, ?string $throw = '$input', ?array &$typesChecked = null): ?string
```

Reply:
I understand that the method is very powerful, but at the same time it is also very complex. 

For example, when I see the following call somewhere in the code, I am completely confused:

```php
$someString = DataTypeHelper::checkType($value, ['array', 'string'], false, '$value', &$typesChecked);
```

I would be completely fine with a method like, which throws an exception on failure
```php
DataTypeHelper::ensureType(mixed $value, array $types) : void;
``` 

A few other thoughts/ambiguities for `checkType()` which is why I want to keep it `private`:

- Why `?string` as return type? I would expect `bool` for such type of helper
- Do we really need `$requireAll` option? I would drop it for simplicity.
- I understand the `$throw` is useful in some situations, but I don't like the style to pass a variable name as string. 
- I roughly remember that you explained the purpose of `&$typesChecked`
- Before we introduce a more complex type validation/helper API (500 lines+ ), we should perhaps also check whether there are ready-made libraries for this. We can very easily replace helper methods `filterInt()` with other implementations later on. However, this will not be possible with more complex methods such as `checkType()`.

Reply:
> > I would be completely fine with a method like, which throws an exception on failure
> > ```
> > DataTypeHelper::ensureType(mixed $value, array $types) : void;
> > ```
> 
> How about this:
> 
> * One "ensure" method (throwing an exception)
>   ```
>   DataTypeHelper::ensureType($input, array $types, string $parameterOrMessage = '$input'): void
>   ```
> * One "check" method (returning the type)
>   ```
>   DataTypeHelper::checkType($input, array $types, bool $returnIndex = false): ?string
>   ```
>
> > Why `?string` as return type? I would expect `bool` for such type of helper
> 
> Because like that you can use it e.g. in a `switch` statement like so:
> 
> ```
> switch (DataTypeHelper::checkType($value, ['string', 'int', Stringable::class])) {
>     case Stringable::class:
>         $value = (string)$value;
>     case 'string':
>         // Do something with a string value
>         break;
>         
>     case 'int':
>         // do something else with an int
>         break;
>         
>     default:
>         // do nothing
> }
> ```
>

Ok, I see. How about then calling the method `DataTypeHelper::getType(mixed $mixed, array $allowedTypes): ?string`?

So:

- `DataTypeHelper::ensureType(mixed $input, array $allowedTypes): void`
- `DataTypeHelper::getType(mixed $input, array $allowedTypes): ?string` 
- `DataTypeHelper::hasType(mixed $input, array $allowedTypes): bool` 

---

I'm still not sure, about:

- Passing the variable name as string to the `ensureType()` method. For me the code usage looks cluttered. `if (DataTypeHelper::ensureType($param1, ['string'], '$param1') && DataTypeHelper::ensureType($param2, ['int'], '$param2')) { /* ... */ }`
- I'm thinking about raising the HH minVersion to `8.1` and use Enums instead of strings

Reply:
> Ok, I see. How about then calling the method `DataTypeHelper::getType(mixed $mixed, array $allowedTypes): ?string`?
> 
> So:
> 
> * `DataTypeHelper::ensureType(mixed $input, array $allowedTypes): void`
> * `DataTypeHelper::getType(mixed $input, array $allowedTypes): ?string`
> * `DataTypeHelper::hasType(mixed $input, array $allowedTypes): bool`

Sure, that could be done.

However, `hasType` is kinda just an alias to `getType`: except for the strict type, the later would always yield the same result as the former, since only the empty string `""` would yield `false`, but that could never happen. In such a case `null` would be returned which is `==` to `false`:

```php
if (DataTypeHelper::getType("some string", ['string', 'int'])) {
    // result "string" is "true"
}
```

The only time you would actually need to use `DataTypeHelper::hasType()` (with a strict boolean return type) is when you'd pass the result straight to another fixed-type method:

```php
function foo(bool $isString) {
   // $isString cannot be auto-converted here.
}

// so this would fail:
foo(DataTypeHelper::getType("some string", ['string', 'int'])));

// instead, this would be required:
DataTypeHelper::getType("some string", ['string', 'int']) !== null); 
```

As such, that would also be the exact implementation:
```php
public function hasType($input, array $allowedTypes): bool {
    return DataTypeHelper::getType($input, $allowedTypes) !== null);
}
```

> 
> I'm still not sure, about:
> 
> * Passing the variable name as string to the `ensureType()` method. For me the code usage looks cluttered. `if (DataTypeHelper::ensureType($param1, ['string'], '$param1') && DataTypeHelper::ensureType($param2, ['int'], '$param2')) { /* ... */ }`

I understand. Well, in such a case, where you only have one possible type, you could use a strictly-typed parameter, which is always the better option anyway. (But I don't think that was the point of your argument.)

Furthermore you would no longer need the `if`. And rewriting it without the `&&` would certainly increase readability:

```php
function bar($param1, $param2) {
    DataTypeHelper::ensureType($param1, ['string', 'int'], '$param1');
    DataTypeHelper::ensureType($param2, [someClass::class, 'is_scalar'], 'Please make sure $param2 has the right format');

    /* continue with type-safe values here */
}
```

Besides, the parameter is optional. But it helps to write a meaningful exception message. Which also is as close to the PHP's own type exception (naming the parameter name and number).

If you do not want to throw an exception, but just do something if a given type is povided, then you do no longer have that parameter:

```php
if (DataTypeHelper::hasType($param1, ['string', 'int']) && DataTypeHelper::hasType($param2, [someClass::class, 'is_scalar']) {
     /* do something conditionally with type-safe values here */
}
```


> * I'm thinking about raising the HH minVersion to `8.1` and use Enums instead of strings

You mean to raise the PHP min version?

I'm not quite sure what would enum's help in this context? For the `$allowedTypes` parameter?

Reply:
> > Ok, I see. How about then calling the method `DataTypeHelper::getType(mixed $mixed, array $allowedTypes): ?string`?
> > So:
> > 
> > * `DataTypeHelper::ensureType(mixed $input, array $allowedTypes): void`
> > * `DataTypeHelper::getType(mixed $input, array $allowedTypes): ?string`
> > * `DataTypeHelper::hasType(mixed $input, array $allowedTypes): bool`
> 
> Sure, that could be done.
> 
> However, `hasType` is kinda just an alias to `getType`: except for the strict type, the later would always yield the same result as the former, since only the empty string `""` would yield `false`, but that could never happen. In such a case `null` would be returned which is `==` to `false`:
> 
> ```
> if (DataTypeHelper::getType("some string", ['string', 'int'])) {
>     // result "string" is "true"
> }
> ```
> 
> The only time you would actually need to use `DataTypeHelper::hasType()` (with a strict boolean return type) is when you'd pass the result straight to another fixed-type method:
> 
> ```
> function foo(bool $isString) {
>    // $isString cannot be auto-converted here.
> }
> 
> // so this would fail:
> foo(DataTypeHelper::getType("some string", ['string', 'int'])));
> 
> // instead, this would be required:
> DataTypeHelper::getType("some string", ['string', 'int']) !== null); 
> ```
> 
> As such, that would also be the exact implementation:
> 
> ```
> public function hasType($input, array $allowedTypes): bool {
>     return DataTypeHelper::getType($input, $allowedTypes) !== null);
> }
> ```
> 
> > I'm still not sure, about:
> > 
> > * Passing the variable name as string to the `ensureType()` method. For me the code usage looks cluttered. `if (DataTypeHelper::ensureType($param1, ['string'], '$param1') && DataTypeHelper::ensureType($param2, ['int'], '$param2')) { /* ... */ }`
> 
> I understand. Well, in such a case, where you only have one possible type, you could use a strictly-typed parameter, which is always the better option anyway. (But I don't think that was the point of your argument.)
> 
> Furthermore you would no longer need the `if`. And rewriting it without the `&&` would certainly increase readability:
> 
> ```
> function bar($param1, $param2) {
>     DataTypeHelper::ensureType($param1, ['string', 'int'], '$param1');
>     DataTypeHelper::ensureType($param2, [someClass::class, 'is_scalar'], 'Please make sure $param2 has the right format');
> 
>     /* continue with type-safe values here */
> }
> ```
> 
> Besides, the parameter is optional. But it helps to write a meaningful exception message. Which also is as close to the PHP's own type exception (naming the parameter name and number).
> 
> If you do not want to throw an exception, but just do something if a given type is povided, then you do no longer have that parameter:
> 
> ```
> if (DataTypeHelper::hasType($param1, ['string', 'int']) && DataTypeHelper::hasType($param2, [someClass::class, 'is_scalar']) {
>      /* do something conditionally with type-safe values here */
> }
> ```
> 
> > * I'm thinking about raising the HH minVersion to `8.1` and use Enums instead of strings
> 
> You mean to raise the PHP min version?
> 
> I'm not quite sure what would enum's help in this context? For the `$allowedTypes` parameter?

For me, I would prefer instead of `if (static::getType($var, ['string', 'int']) === 'string') { ... }`  something like: `if (static::getType($var, [DataType::String, DataType::Integer]) === DataType::String) { ... }`.

Reply:
> > > * I'm thinking about raising the HH minVersion to `8.1` and use Enums instead of strings
> > 
> > 
> > You mean to raise the PHP min version?
> > I'm not quite sure what would enum's help in this context? For the `$allowedTypes` parameter?
> 
> For me, I would prefer instead of `if (static::getType($var, ['string', 'int']) === 'string') { ... }` something like: `if (static::getType($var, [DataType::String, DataType::Integer]) === DataType::String) { ... }`.

I see.

Well, if you raise PHP to v8.1, then I'm happy to create such an Enum.

Otherwise, I'm happy to create regular constants in the current `DataTypeHelper` class.